### PR TITLE
Optimize kubelet stats provider for perfomace bottleneck

### LINKED ISSUE #103423: Kubelet stats Provider performance bottleneck
For single node with hundreds and thousands of pods and containers, the cpu usage of kubelet becomes high, one bottleneck is the cadvisor stats provider as below:
https://github.com/kubernetes/kubernetes/blob/659c7e709f3b7f5f2a25e456525cd8747f2e68cc/pkg/kubelet/stats/cadvisor_stats_provider.go#L156

```
File: kubelet
Build ID: 9c138a005862ce370399990317dffd36e56def0b
Type: cpu
Time: Jul 1, 2021 at 3:06pm (CST)
Duration: 30.18s, Total samples = 1.36mins (270.50%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 39.64s, 48.55% of 81.65s total
Dropped 932 nodes (cum <= 0.41s)
Showing top 10 nodes out of 303
      flat  flat%   sum%        cum   cum%
     9.10s 11.15% 11.15%      9.30s 11.39%  syscall.Syscall6
     6.60s  8.08% 19.23%      6.75s  8.27%  syscall.Syscall
     5.55s  6.80% 26.03%     10.44s 12.79%  runtime.scanobject
     3.19s  3.91% 29.93%      3.53s  4.32%  runtime.findObject
     3.01s  3.69% 33.62%      3.01s  3.69%  strings.LastIndexByte
     2.90s  3.55% 37.17%      8.30s 10.17%  runtime.mallocgc
     2.67s  3.27% 40.44%      3.81s  4.67%  runtime.mapiternext
     2.41s  2.95% 43.39%      2.41s  2.95%  runtime.duffcopy
     2.15s  2.63% 46.03%      2.15s  2.63%  memeqbody
     2.06s  2.52% 48.55%     18.19s 22.28%  k8s.io/kubernetes/pkg/kubelet/stats.
getCadvisorPodInfoFromPodUID
```
the `getCadvisorPodInfoFromPodUID` method iterate all containerInfos for every podStats cause high cpu 

#### What you expected to happen:

the `getCadvisorPodInfoFromPodUID` method should not iterate allcontainerInfos for every podStats,  use map instead

#### Environment:
- Kubernetes version (use `kubectl version`): v1.19.4
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by 249043822**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #103423

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 1]
```diff
@@ -91,9 +91,8 @@ func (p *cadvisorStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to get container info from cadvisor: %v", err)
 	}
-	// removeTerminatedContainerInfo will also remove pod level cgroups, so save the infos into allInfos first
-	allInfos := infos
-	infos = removeTerminatedContainerInfo(infos)
+
```
Comment (haircommander) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:94]: I think these could use more informative names now that we've lost this comment:
`filteredInfos, allInfos` perhaps?

**Comment by haircommander** [comment]
one nit, otherwise LGTM

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 2]
```diff
@@ -303,20 +300,58 @@ func isPodManagedContainer(cinfo *cadvisorapiv2.ContainerInfo) bool {
...
+	}
+	return nil
+}
+
+// filterTerminatedContainerInfoAndAssembleByPodCgroupKey returns the specified containerInfo but with
+// the stats of the terminated containers removed and all containerInfos assembled by pod cgroup key.
+//
+// A ContainerInfo is considered to be of a terminated container if it has an
+// older CreationTime and zero CPU instantaneous and memory RSS usage.
+func filterTerminatedContainerInfoAndAssembleByPodCgroupKey(containerInfo map[string]cadvisorapiv2.ContainerInfo) (map[string]cadvisorapiv2.ContainerInfo, map[string]cadvisorapiv2.ContainerInfo) {
```
Comment (bobbypage) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:314]: I think this replaces `removeTerminatedContainerInfo` now and I don't see any more usages of `removeTerminatedContainerInfo`. Can we delete it then?

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 3]
```diff
@@ -303,20 +300,58 @@ func isPodManagedContainer(cinfo *cadvisorapiv2.ContainerInfo) bool {
...
 func getCadvisorPodInfoFromPodUID(podUID types.UID, infos map[string]cadvisorapiv2.ContainerInfo) *cadvisorapiv2.ContainerInfo {
-	for key, info := range infos {
+	if info, found := infos[cm.GetPodCgroupNameSuffix(podUID)]; found {
+		return &info
+	}
+	return nil
+}
+
+// filterTerminatedContainerInfoAndAssembleByPodCgroupKey returns the specified containerInfo but with
+// the stats of the terminated containers removed and all containerInfos assembled by pod cgroup key.
```
Comment (bobbypage) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:310]: can we make explicit in the godoc for the two maps returns what is the key -> value mapping for each of the two maps?

**Comment by 249043822** [comment]
@haircommander @bobbypage @ehashman Thanks for your review, I fixed all the nits

**Comment by haircommander** [comment]
LGTM, I would consider this a worthy bugfix for 1.22
/milestone v1.22

**Comment by bobbypage** [comment]
thanks!

/lgtm

**Comment by ehashman** [comment]
/remove-priority backlog
/priority important-soon

Impact: Stats gathering via cadvisor is expensive and consumes a lot of CPU on a node running many containers. https://github.com/kubernetes/kubernetes/issues/103423#issue-935467574

Not a regression, present since at least 1.19.4. Possibly a good candidate for backport. Not possible to unit test, hence no test changes.

Will refer to RT to make the call re: inclusion since we have passed code freeze. We could either take this now, and backport after soaking, or wait until 1.23 opens, soak for a month or so, and backport.

**Review by mrunalp — APPROVED** [review]

**Pull request merged**